src/v/cloud_io/basic_cache_service_api.h

```diff
@@ -0,0 +1,159 @@
+#include <seastar/core/iostream.hh>
+#include <seastar/core/lowres_clock.hh>
+
+#include <filesystem>
+#include <iterator>
+#include <optional>
+#include <set>
+#include <string_view>
+
+namespace cloud_io {
+
+static constexpr size_t default_write_buffer_size = 128_KiB;
+static constexpr unsigned default_write_behind = 10;
+static constexpr size_t default_read_buffer_size = 128_KiB;
+static constexpr unsigned default_read_ahead = 10;
+
+struct [[nodiscard]] cache_item {
+    ss::file body;
+    size_t size;
+};
+
+struct [[nodiscard]] cache_item_str {
+    ss::input_stream<char> body;
+    size_t size;
+};
+
+enum class [[nodiscard]] cache_element_status {
+    available,
+    not_available,
+    in_progress,
+};
+
+std::ostream& operator<<(std::ostream& o, cache_element_status);
+
+template<class Clock>
+class basic_cache_service_api;
+
+/// RAII guard for bytes reserved in the cache: constructed prior to a call
+/// to cache::put, and may be destroyed afterwards.
+template<class Clock>
+class basic_space_reservation_guard {
+public:
+    basic_space_reservation_guard(
+      basic_cache_service_api<Clock>& cache,
+      uint64_t bytes,
+      size_t objects) noexcept
+      : _cache(cache)
+      , _bytes(bytes)
+      , _objects(objects) {}
+
+    basic_space_reservation_guard(const basic_space_reservation_guard&)
+      = delete;
+    basic_space_reservation_guard() = delete;
+    basic_space_reservation_guard(basic_space_reservation_guard&& rhs) noexcept
+      : _cache(rhs._cache)
+      , _bytes(rhs._bytes)
+      , _objects(rhs._objects) {
+        rhs._bytes = 0;
+        rhs._objects = 0;
+    }
+
+    ~basic_space_reservation_guard();
+
+    /// After completing the write operation that this space reservation
+    /// protected, indicate how many bytes were really written: this is used to
+    /// atomically update cache usage stats to free the reservation and update
+    /// the bytes used stats together.
+    ///
+    /// May only be called once per reservation.
+    void wrote_data(uint64_t, size_t);
+
+private:
+    basic_cache_service_api<Clock>& _cache;
+
+    // Size acquired at time of reservation
+    uint64_t _bytes{0};
+    size_t _objects{0};
+};
+
+template<class Clock = ss::lowres_clock>
+class basic_cache_service_api {
+public:
+    basic_cache_service_api() = default;
+    basic_cache_service_api(const basic_cache_service_api&) = default;
+    basic_cache_service_api& operator=(const basic_cache_service_api&)
+      = default;
+    basic_cache_service_api(basic_cache_service_api&&) = default;
+    basic_cache_service_api& operator=(basic_cache_service_api&&) = default;
+    virtual ~basic_cache_service_api() = default;
+
+    /// Get cached value as a stream if it exists
+    ///
+    /// \param key is a cache key
+    /// \param read_buffer_size is a read buffer size for the iostream
+    /// \param readahead number of pages that can be read asynchronously
+    virtual ss::future<std::optional<cache_item_str>> get(
```

Comment:
i'm reading the return type as `cache item as string` 🙃

Reply:
same. i think we can rename this later

Reply:
https://github.com/redpanda-data/redpanda/pull/23810